go/store/prolly/tuple_range.go

```diff
@@ -167,18 +173,161 @@ func (r Range) Matches(t val.Tuple) bool {
+		if !r.Fields[i].BoundsAreEqual {
 			return false
 		}
 	}
 	return true
 }
 
+// KeyRangeLookup will return a stop key and true if the range can be scanned
+// from a start to stop tuple. Otherwise, return a nil key and false. A range
+// can be key range scanned if the prefix is exact, and the final field is
+// numeric or string. The stop key adds +1 to a numeric final field, and appends
+// '0' to a string final field.
+// TODO: support non-exact final field, and use range upper bound?
+func (r Range) KeyRangeLookup(pool pool.BuffPool) (val.Tuple, bool) {
+	if r.Tup == nil {
+		return nil, false
+	}
+	n := len(r.Fields) - 1
+	for i := range r.Fields {
+		if r.Fields[i].Lo.Value == nil {
+			if r.Fields[i].Hi.Value != nil {
+				return nil, false
+			}
+			n = i - 1
+			break
+		}
+		if !r.Fields[i].BoundsAreEqual {
+			return nil, false
+		}
+	}
+
+	if n < 0 {
+		// ex: range scan
+		return nil, false
+	}
+
+	for _, typ := range r.Desc.Types[n+1:] {
+		if !typ.Nullable {
+			// this is checked separately because fulltext descriptors
+			// do not match field lengths sometimes
+			// todo: why?
+			return nil, false
+		}
+	}
+
+	for i := n + 1; i < len(r.Fields); i++ {
+		if r.Fields[i].Lo.Value != nil ||
+			r.Fields[i].Hi.Value != nil {
+			// these shouldn't be possible with regular index semantics,
+			// but we manually inline indexes sometimes on the Dolt side,
+			// and it's possible we'd change analyzer indexing semantics
+			// in the future
+			return nil, false
+		}
+
+	}
+
+	tb := val.NewTupleBuilder(r.Desc)
+	for i := 0; i < n; i++ {
+		if i != n {
+			// direct copy all but the last field
+			tb.PutRaw(i, r.Tup.GetField(i))
+		}
+	}
+
+	// last field will be incremented by one to get the exclusive key
+	// range [key, key+1)
+	switch r.Desc.Types[n].Enc {
+	case val.StringEnc:
+		v := r.Fields[n].Lo.Value
+		tb.PutString(n, string(v)+"0")
+	case val.Int8Enc:
+		v, ok := r.Desc.GetInt8(n, r.Tup)
+		if !ok {
+			return nil, false
+		}
+		tb.PutInt8(n, v+1)
+	case val.Uint8Enc:
+		v, ok := r.Desc.GetUint8(n, r.Tup)
+		if !ok {
+			return nil, false
+		}
+		tb.PutUint8(n, v+1)
+	case val.Int16Enc:
+		v, ok := r.Desc.GetInt16(n, r.Tup)
+		if !ok {
+			return nil, false
+		}
+		tb.PutInt16(n, v+1)
+	case val.Uint16Enc:
+		v, ok := r.Desc.GetUint16(n, r.Tup)
+		if !ok {
+			return nil, false
+		}
+		tb.PutUint16(n, v+1)
+	case val.Int32Enc:
+		v, ok := r.Desc.GetInt32(n, r.Tup)
+		if !ok {
+			return nil, false
+		}
+		tb.PutInt32(n, v+1)
+	case val.Uint32Enc:
+		v, ok := r.Desc.GetUint32(n, r.Tup)
+		if !ok {
+			return nil, false
+		}
+		tb.PutUint32(n, v+1)
+	case val.Int64Enc:
+		v, ok := r.Desc.GetInt64(n, r.Tup)
+		if !ok {
+			return nil, false
+		}
+		tb.PutInt64(n, v+1)
+	case val.Uint64Enc:
+		v, ok := r.Desc.GetUint64(n, r.Tup)
+		if !ok {
+			return nil, false
+		}
+		tb.PutUint64(n, v+1)
+	case val.Float32Enc:
+		v, ok := r.Desc.GetFloat32(n, r.Tup)
+		if !ok {
+			return nil, false
+		}
+		tb.PutFloat32(n, v+1)
```

Comment:
Instead of adding 1 here (and elsewhere), we could find the next highest representable float. This would be useful both for very high values that don't have enough precision to add 1, but also very low values where adding 1 is an enormous range.